appium-tests/helpers/contactsHelper.js

```diff
@@ -0,0 +1,222 @@
+
+    var result =
+        'var results = document.getElementById("info");\n' +
+        'var obj = new ContactFindOptions();\n' +
+        'obj.filter = ' + preparedOldDisplayName + ';\n' +
+        'obj.multiple = false;\n' +
+
+        'navigator.contacts.find(["displayName", "name"], function(contacts) {\n' +
+        '    if (contacts.length === 0) {\n' +
+        '        results.innerHTML = "No contacts to update.";\n' +
+        '        return;\n' +
+        '    }\n' +
+        '    var contact = contacts[0];\n' +
+        '    contact.displayName = ' + preparedNewDisplayName + ';\n' +
+        '    contact.name = ' + preparedNewName + ';\n' +
+        '    contact.save(function(updated) {\n' +
+        '        results.innerHTML = "Contact updated.";\n' +
+        '    },function(e) {\n' +
+        '        results.innerHTML = "Update failed: error " + e.code;\n' +
+        '    });\n' +
+        '}, function(e) {\n' +
+        '    results.innerHTML = "Search failed: error " + e.code;\n' +
+        '}, obj);';
+
+    return result;
+};
+
+module.exports.getRemoveTestContactsCode = function () {
+    var result =
+        'var results = document.getElementById("info");\n' +
+        'results.innerHTML = "";\n' +
+        'var obj = new ContactFindOptions();\n' +
+        'obj.filter = "DeleteMe";\n' +
+        'obj.multiple = true;\n' +
+        'navigator.contacts.find(["note"], function(contacts) {\n' +
+        '    var removes = [];\n' +
+        '    contacts.forEach(function(contact) {\n' +
+        '        removes.push(contact);\n' +
+        '    });\n' +
+        '    if (removes.length === 0) {\n' +
+        '        results.innerHTML = "No contacts to remove";\n' +
+        '        return;\n' +
+        '    }\n' +
+
+        '   var nextToRemove;\n' +
+        '   if (removes.length > 0) {\n' +
+        '        nextToRemove = removes.shift();\n' +
+        '    }\n' +
+
+        '    function removeNext(item) {\n' +
+        '        if (typeof item === "undefined") {\n' +
+        '            return;\n' +
+        '        }\n' +
+
+        '        if (removes.length > 0) {\n' +
+        '            nextToRemove = removes.shift();\n' +
+        '        } else {\n' +
+        '            nextToRemove = undefined;\n' +
+        '        }\n' +
+
+        '        item.remove(function removeSucceeded() {\n' +
+        '            results.innerHTML += "Removed a contact with ID " + item.id + "<br/>";\n' +
+        '            removeNext(nextToRemove);\n' +
+        '        }, function removeFailed() {\n' +
+        '            results.innerHTML += "Failed to remove a contact with ID " + item.id + "<br/>";\n' +
+        '            removeNext(nextToRemove);\n' +
+        '        });\n' +
+        '    }\n' +
+        '    removeNext(nextToRemove);\n' +
+        '}, function(e) {\n' +
+        '   results.innerHTML = "Search failed: error " + e.code;\n' +
+        '}, obj);';
+
+    return result;
+};
```

Comment:
There are other way: we could just click the buttons in the mobilespec app's manual tests.
I've found a couple of downsides of such approach:
- We are tightly tied to the mobilespec app and its implementation: should there be changes to it's markup/navigation, or to the manual tests, we're bound to change the appium tests too
- Appium tests look more complicated. For example, to debug a failing Appium test which clicks a button in the manual test I need to look at the Appium code, see which button is clicked, like here: `elementByXPath('//a[text()="Add new nameless contact"]')`, go to the manual tests, find the code that creates this particular button and finally find a click handler function for this particular button. All this opposed to just having the executed code close at hand, in the helper function.
- Clicking buttons can be difficult :) For example, Appium is having a huge problem scrolling the page to find a button which is out of view on iOS simulator.

Could you please describe your thoughts about the current implementation and why are you against it?